File: python/tests/guardrails_response_validator_test.py

from pedro_agentware.middleware.guardrails.nudge import NudgeKind
from pedro_agentware.middleware.guardrails.response_validator import (
    ResponseValidator,
    ToolCall,
)


class TestResponseValidator:
    """Tests for ResponseValidator."""

    def test_new_response_validator(self):
        """Test creating a new ResponseValidator."""
        tools = ["tool1", "tool2"]
        rv = ResponseValidator(tools, rescue_enabled=True)

        assert len(rv.tool_names) == 2
        assert rv.rescue_enabled is True

    def test_validate_tool_calls_valid(self):
        """Test validating valid tool calls."""
        tools = ["tool1", "tool2"]
        rv = ResponseValidator(tools, rescue_enabled=False)

        calls = [ToolCall(tool="tool1", args={"key": "value"})]
        result = rv.validate_tool_calls(calls)

        assert result.needs_retry is False
        assert result.nudge is None
        assert len(result.tool_calls) == 1

    def test_validate_tool_calls_invalid(self):
        """Test validating invalid tool calls."""
        tools = ["tool1", "tool2"]
        rv = ResponseValidator(tools, rescue_enabled=False)

        calls = [ToolCall(tool="unknown_tool", args={})]
        result = rv.validate_tool_calls(calls)

        assert result.needs_retry is True
        assert result.nudge is not None
        assert result.nudge.kind == NudgeKind.UNKNOWN_TOOL

    def test_validate_text_response_with_rescue(self):
        """Test rescuing a text response into tool calls."""
        tools = ["tool1", "tool2"]
        rv = ResponseValidator(tools, rescue_enabled=True)

        response = '{"tool": "tool1", "args": {"key": "value"}}'
        result = rv.validate_text_response(response)

        assert result.needs_retry is False
        assert len(result.tool_calls) == 1

    def test_validate_text_response_without_rescue(self):
        """Test text response without rescue enabled."""
        tools = ["tool1", "tool2"]
        rv = ResponseValidator(tools, rescue_enabled=False)

        response = "This is just text."
        result = rv.validate_text_response(response)

        assert result.needs_retry is True
        assert result.nudge is not None
        assert result.nudge.kind == NudgeKind.RETRY

    def test_validate_text_response_empty(self):
        """Test empty text response."""
        tools = ["tool1", "tool2"]
        rv = ResponseValidator(tools, rescue_enabled=True)

        result = rv.validate_text_response("")

        assert result.needs_retry is True

    def test_rescue_tool_call_strips_think_tags(self):
        """Test that think tags are stripped."""
        tools = ["tool1", "tool2"]
        rv = ResponseValidator(tools, rescue_enabled=True)

        response = "[THINK] thinking [/THINK]\n{\"tool\": \"tool1\", \"args\": {}}"
        calls = rv._rescue_tool_call(response)

        assert len(calls) == 1

    def test_rescue_tool_call_strips_python_tag(self):
        """Test that python tags are stripped."""
        tools = ["tool1", "tool2"]
        rv = ResponseValidator(tools, rescue_enabled=True)

        response = "<|python_tag|>{\"tool\": \"tool1\", \"args\": {}}"
        calls = rv._rescue_tool_call(response)

        assert len(calls) == 1

    def test_extract_json_tool_calls_with_code_fence(self):
        """Test extracting JSON from code fences."""
        tools = ["tool1", "tool2"]
        rv = ResponseValidator(tools, rescue_enabled=True)

        response = '```json\n{"tool": "tool1", "args": {"key": "value"}}\n```'
        calls = rv._extract_json_tool_calls(response)

        assert len(calls) == 1

    def test_extract_rehearsal_tool_calls(self):
        """Test extracting rehearsal format tool calls."""
        tools = ["tool1", "tool2"]
        rv = ResponseValidator(tools, rescue_enabled=True)

        response = 'tool1[ARGS]{"key": "value"}'
        calls = rv._extract_rehearsal_tool_calls(response)

        assert len(calls) == 1

    def test_try_parse_tool_call_valid(self):
        """Test parsing valid JSON tool call."""
        tools = ["tool1", "tool2"]
        rv = ResponseValidator(tools, rescue_enabled=True)

        json_str = '{"tool": "tool1", "args": {"key": "value"}}'
        call = rv._try_parse_tool_call(json_str)

        assert call is not None
        assert call.tool == "tool1"

    def test_try_parse_tool_call_invalid_json(self):
        """Test parsing invalid JSON."""
        tools = ["tool1", "tool2"]
        rv = ResponseValidator(tools, rescue_enabled=True)

        call = rv._try_parse_tool_call("not valid json")

        assert call is None

    def test_try_parse_tool_call_unknown_tool(self):
        """Test parsing with unknown tool."""
        tools = ["tool1"]
        rv = ResponseValidator(tools, rescue_enabled=True)

        json_str = '{"tool": "unknown", "args": {}}'
        call = rv._try_parse_tool_call(json_str)

        assert call is None

    def test_extract_qwen_xml_tool_calls(self):
        """Test Qwen XML extraction (not implemented)."""
        tools = ["tool1", "tool2"]
        rv = ResponseValidator(tools, rescue_enabled=True)

        calls = rv._extract_qwen_xml_tool_calls("<function=tool1>content</function>")

        assert len(calls) == 0

File: typescript/tests/guardrails-response-validator.test.ts
import { ResponseValidator } from "../src/middleware/guardrails/response_validator";
import { NudgeKind } from "../src/middleware/guardrails/nudge";

describe("ResponseValidator", () => {
  describe("new ResponseValidator", () => {
    it("should create a new validator", () => {
      const rv = new ResponseValidator(["tool1", "tool2"], true);
      expect(rv).toBeDefined();
    });
  });

  describe("validateToolCalls", () => {
    it("should validate valid tool calls", () => {
      const rv = new ResponseValidator(["tool1", "tool2"], false);
      const calls = [{ tool: "tool1", args: { key: "value" } }];
      const result = rv.validateToolCalls(calls);

      expect(result.needsRetry).toBe(false);
      expect(result.nudge).toBeNull();
      expect(result.toolCalls.length).toBe(1);
    });

    it("should reject unknown tools", () => {
      const rv = new ResponseValidator(["tool1", "tool2"], false);
      const calls = [{ tool: "unknown_tool", args: {} }];
      const result = rv.validateToolCalls(calls);

      expect(result.needsRetry).toBe(true);
      expect(result.nudge).not.toBeNull();
      expect(result.nudge!.kind).toBe(NudgeKind.UNKNOWN_TOOL);
    });
  });

  describe("validateTextResponse", () => {
    it("should rescue tool calls from text", () => {
      const rv = new ResponseValidator(["tool1", "tool2"], true);
      const response = '{"tool": "tool1", "args": {"key": "value"}}';
      const result = rv.validateTextResponse(response);

      expect(result.needsRetry).toBe(false);
      expect(result.toolCalls.length).toBe(1);
    });

    it("should return nudge for text without rescue", () => {
      const rv = new ResponseValidator(["tool1", "tool2"], false);
      const response = "This is just text.";
      const result = rv.validateTextResponse(response);

      expect(result.needsRetry).toBe(true);
      expect(result.nudge).not.toBeNull();
      expect(result.nudge!.kind).toBe(NudgeKind.RETRY);
    });

    it("should return nudge for empty response", () => {
      const rv = new ResponseValidator(["tool1", "tool2"], true);
      const result = rv.validateTextResponse("");

      expect(result.needsRetry).toBe(true);
    });
  });

  describe("rescueToolCall", () => {
    it("should strip think tags", () => {
      const rv = new ResponseValidator(["tool1", "tool2"], true);
      const response = "[THINK] thinking [/THINK]\n{\"tool\": \"tool1\", \"args\": {}}";

      // Access private method through any cast
      const calls = (rv as unknown as { rescueToolCall: (r: string) => { tool: string; args: Record<string, unknown> }[] }).rescueToolCall(response);

      expect(calls.length).toBe(1);
    });

    it("should strip python tag", () => {
      const rv = new ResponseValidator(["tool1", "tool2"], true);
      const response = "<|python_tag|>{\"tool\": \"tool1\", \"args\": {}}";

      const calls = (rv as unknown as { rescueToolCall: (r: string) => { tool: string; args: Record<string, unknown> }[] }).rescueToolCall(response);

      expect(calls.length).toBe(1);
    });
  });

  describe("extractJsonToolCalls", () => {
    it("should extract from code fence", () => {
      const rv = new ResponseValidator(["tool1", "tool2"], true);
      const response = '```json\n{"tool": "tool1", "args": {"key": "value"}}\n```';

      const calls = (rv as unknown as { extractJsonToolCalls: (r: string) => { tool: string; args: Record<string, unknown> }[] }).extractJsonToolCalls(response);

      expect(calls.length).toBe(1);
    });
  });

  describe("extractRehearsalToolCalls", () => {
    it("should extract rehearsal format", () => {
      const rv = new ResponseValidator(["tool1", "tool2"], true);
      const response = 'tool1[ARGS]{"key": "value"}';

      const calls = (rv as unknown as { extractRehearsalToolCalls: (r: string) => { tool: string; args: Record<string, unknown> }[] }).extractRehearsalToolCalls(response);

      expect(calls.length).toBe(1);
    });
  });

  describe("tryParseToolCall", () => {
    it("should parse valid JSON", () => {
      const rv = new ResponseValidator(["tool1", "tool2"], true);
      const jsonStr = '{"tool": "tool1", "args": {"key": "value"}}';

      const call = (rv as unknown as { tryParseToolCall: (j: string) => { tool: string; args: Record<string, unknown> } | null }).tryParseToolCall(jsonStr);

      expect(call).not.toBeNull();
      expect(call!.tool).toBe("tool1");
    });

    it("should return null for invalid JSON", () => {
      const rv = new ResponseValidator(["tool1", "tool2"], true);
      const call = (rv as unknown as { tryParseToolCall: (j: string) => { tool: string; args: Record<string, unknown> } | null }).tryParseToolCall("not valid json");

      expect(call).toBeNull();
    });

    it("should return null for unknown tool", () => {
      const rv = new ResponseValidator(["tool1"], true);
      const jsonStr = '{"tool": "unknown", "args": {}}';

      const call = (rv as unknown as { tryParseToolCall: (j: string) => { tool: string; args: Record<string, unknown> } | null }).tryParseToolCall(jsonStr);

      expect(call).toBeNull();
    });
  });

  describe("extractQwenXmlToolCalls", () => {
    it("should return empty (not implemented)", () => {
      const rv = new ResponseValidator(["tool1", "tool2"], true);
      const response = "<function=tool1>content</function>";

      const calls = (rv as unknown as { extractQwenXmlToolCalls: (r: string) => { tool: string; args: Record<string, unknown> }[] }).extractQwenXmlToolCalls(response);

      expect(calls.length).toBe(0);
    });
  });
});